### ISSUE TITLE
[VRL] `truncate()` inconsistency: operates on UTF-8 characters instead of bytes

### ISSUE THREAD

**Issue by hhromic** [vectordotdev/vector#12369]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

I'm not sure if this is a bug or a feature request. I tend to consider it a bug due to my own expectations.

String manipulation functions in VRL seem to mostly operate on bytes and not UTF-8 characters (unicode scalars).

This was recently brought up for the `length()` function in https://github.com/vectordotdev/vector/issues/12019, where a new `strlen()` was implemented to operate on UTF-8 characters instead of bytes.

However, I just noticed that the `truncate()` function is operating in UTF-8 characters instead of bytes, creating an inconstency.

This is problematic for us because we use `truncate()` to limit the number of bytes in a field that later is used for insertion into a `VARCHAR(size)` column in an SQL database, which defines `size` in bytes (highly likely an SQL standard).

Consider the following REPL examples:
```
$ truncate(s'hello', 2)
"he"

$ length(truncate(s'hello', 2))
2

$ truncate(s'ñandú', 2)  # expected just 'ñ' here due to using 2 bytes
"ña"

$ length(truncate(s'ñandú', 2))  # expected 2 here
3
```

For completeness, I checked up the `slice()` function and indeed operates on bytes as expected, like `length()`:
```
$ slice!(s'ñandú', 0)
"ñandú"

$ slice!(s'ñandú', 1)
"�andú"

$ slice!(s'ñandú', 2)
"andú"

$ slice!(s'ñandú', 3)
"ndú"

$ slice!(s'ñandú', 4)
"dú"

$ slice!(s'ñandú', 5)
"ú"

$ slice!(s'ñandú', 6)
"�"

$ slice!(s'ñandú', 7)
""

$ slice!(s'ñandú', 8)
function call error for "slice" at (0:21): "start" must be between "-7" and "7"
```

Fixing this is probably going to be tricky because of the obvious breakage :(

### Version

vector 0.21.0 (x86_64-unknown-linux-gnu c1edb89 2022-04-14)

**Comment by jszwedko** [comment]
Thanks @hhromic . I agree we should be consistent here. It is odd that `length` and `slice` don't match. I could see introducing analogous `strtruncate` and `strslice` methods that operate on characters rather than bytes.

**Comment by hhromic** [comment]
Actually, `length` and `slice` do match, considering that they both work properly on _bytes_.
The one that does not match is `truncate` which currently operates on _characters_.
So, `truncate` is currently what you would call `strtruncate` and a new `truncate` that operates on _bytes_ would be needed.

To better visualise the current situation in VRL:
Function | Uses Bytes | Uses Characters
---|---|---
`truncate` | No | Yes
`length` | Yes | No
`slice` | Yes | No
`strlen` | No | Yes

**Comment by jszwedko** [comment]
Right, yeah, sorry, I meant `truncate` is the odd one out 😄 Considering `length` and `slice` it feels like `truncate` should also operate on bytes.

**Comment by hhromic** [comment]
To help a bit, I went to audit all the functions categorisd as [string functions](https://vector.dev/docs/reference/vrl/functions/#string-functions).

To my surprise, most of them actually operate on UTF-8 characters as opposed to bytes :P

First, let's get some bytes to test. In this case I'm using the UTF-8 character `ñ` as example (`\xc3` + `\xb1`):
```
$ first_byte = decode_base64!("ww==")
"�"

$ length(first_byte)
1

$ second_byte = decode_base64!("sQ==")
"�"

$ length(second_byte)
1

$ both_bytes = s'ñ'
"ñ"

$ length(both_bytes)
2
```

Now, let's see how each of the string functions in VRL behave with those:
```
$ contains(s'ñ', first_byte)  # does not recognise the first byte alone -> operates with characters
false

$ contains(s'ñ', second_byte)  # does not recognise the second byte alone -> operates with characters
false

$ contains(s'ñ', both_bytes)  # recognises both together -> operates with characters
true

$ downcase(s'ÑANDÚ')  # seems to understand UTF-8 characters perfectly
"ñandú"

$ ends_with(s'údnañ', first_byte)  # this result is correct -> seems to operate with bytes or characters
false

$ ends_with(s'údnañ', second_byte)  # this result is incorrect -> operates with characters (confirmed)
false

$ ends_with(s'údnañ', both_bytes)  # this result is correct -> as confirmed in the previous test
true

$ find(s'a ñandú', first_byte)  # found the first byte at index 2 -> seems to operate with bytes
2

$ find(s'a ñandú', second_byte)  # found the second byte at index 3 -> operates with bytes (confirmed)
3

$ find(s'a ñandú', both_bytes) # found the sequence of bytes starting at index 2 -> as confirmed in previous test
2

# skipping join() as I don't think is relevant

# skipping match() and match_any() as they work with regexes

# skipping redact() as it does not work with strings in filters

$ replace(s'ñandú', first_byte, s'N')  # does not find the first byte -> operates with characters
"ñandú"

$ replace(s'ñandú', second_byte, s'N')  # does not find the second byte -> operates with characters
"ñandú"

$ replace(s'ñandú', both_bytes, s'N')  # replaced the complete character -> operates with characters
"Nandú"

$ split(s'ñandú', first_byte)  # does not split by the first byte -> operates with characters
["ñandú"]

$ split(s'ñandú', second_byte)  # does not split by the second byte -> operates with characters
["ñandú"]

$ split(s'ñandú', both_bytes)  # does split by the complete character -> operates with characters
["", "andú"]

$ starts_with(s'ñandú', first_byte)  # this result is correct -> seems to operate with bytes or characters
true

$ starts_with(s'ñandú', second_byte)  # this result is correct -> seems to operate with bytes or characters
false

$ starts_with(s'ñandú', both_bytes)  # this result is correct -> seems to operate with bytes or characters
true

# skipping strip_ansi_escape_codes() as I don't think is relevant

# skipping strip_whitespace() as I don't think is relevant

$ truncate(s'ñandú', 2)  # this one was confirmed in the original issue to operate with characters
"ña"

$ upcase(s'ñandú')  # seems to understand UTF-8 characters perfectly
"ÑANDÚ"
```

In summary, I think I'm opening a pandora's box with this issue as the inconsistencies are way more than expected :(
Notable is the inconsistent case between `ends_with()` and `starts_with()`.
